cpp/src/parquet/exception.h

```diff
@@ -32,13 +32,24 @@
 #define ARROW_UNUSED(x) UNUSED(x)
 #endif
 
+// Parquet exception to Arrow Status
+
 #define PARQUET_CATCH_NOT_OK(s)                    \
   try {                                            \
     (s);                                           \
   } catch (const ::parquet::ParquetException& e) { \
     return ::arrow::Status::IOError(e.what());     \
   }
 
+#define PARQUET_CATCH_AND_RETURN(s)                \
+  try {                                            \
+    return (s);                                    \
+  } catch (const ::parquet::ParquetException& e) { \
```

Comment:
Is it worth explicitly catching ParquetStatusException here (so that `Invalid` statuses don't get recast to `IOError`s?)

Reply:
I'm afraid I don't understand. Here the goal is to transform all Parquet exceptions into Arrow status codes. This is also the same as `PARQUET_CATCH_NOT_OK` above.

Reply:
`PARQUET_THROW_NOT_OK` throws exceptions which contain a full `Status` rather than just an error string. These could be caught explicitly and the original status returned rather than creating a new status:
```suggestion
  } catch (const ::parquet::ParquetStatusException& e) { \
    return e.status(); \
  } catch (const ::parquet::ParquetException& e) { \
```

I think `PARQUET_CATCH_NOT_OK` would also benefit from this pattern

Reply:
Can you open a new JIRA? I think that's a rather independent concern.

Reply:
https://issues.apache.org/jira/browse/ARROW-7332